A34-weighted-round-robin.md

```diff
@@ -0,0 +1,111 @@
+
+### Overview of `weighted_round_robin` policy
+
+`weighted_round_robin` policy is powered by [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) picker. EDF picker maintains a priority queue of `EdfEntry`. On the top of the queue, it’s the entry with lowest `deadline`. `order_offset` is the tie breaker when two entries have same deadline to maintain FIFO order.
+
+Proposed `EdfEntry`
+```
+struct EdfEntry {
+    // primary key for the priority queue. The entry with least deadline is the top of the queue.
+    double deadline;
+    // secondary key for the priority queue. Used as a tiebreaker for same deadline to maintain FIFO order.
+    uint64 order_offset;
+    // `load_balancing_weight` of this endpoint from the latest `ClusterLoadAssignment`.
+    double weight;
+    // Subchannel data structure of this endpoint.
+    Subchannel subchannel;
+}
+```
+
+- On each call to the `Pick`, EDF picker picks the entry `e` on the top of the queue, returns the subchannel associated with the entry. After that, picker updates the `deadline` of `e` to `e.deadline + 1/weight` and either performs a pop and push the entry back to the queue or key increase operation.
+- `weighted_round_robin` updates the entries in EDF priority queue on any change of subchannel `ConnnectiveState` or endpoint `load_balancing_weight`.
+- If all endpoints have the same `load_balancing_weight`, `EDF` picker degenerates to `round_robin` picker. It's easier to reason and consistent with envoy.
+- Endpoints do not have `load_balancing_weight` assigned are discarded.
+- `weighted_round_robin` should always be updated to the lastest `ClusterLoadAssignment`. It's xDS server's responsibility to maintain consistency.
+
+#### EDF picker interface
+```
+/*
+Add new SubChannel to EDF picker.
+
+weighted_round_robin sees a new SubChannel in the most recent ClusterLoadAssignment or ConnectivityState of SubChannel is READY again.
+*/
+Add(SubChannel, weight)
```

Comment:
The intent of our LB policy API is that a picker object should generally be immutable from the outside.  When the routing state changes, the LB policy creates a new picker and gives it to the channel, which will use it to replace the previous picker.  So there's no need for these kinds of mutable methods on the picker.